Client/Frontend/Sync/BraveCore/Bookmarks/BookmarkAPIExtensions.swift

```diff
@@ -0,0 +1,255 @@
+import CoreData
+import Shared
+
+private let log = Logger.browserLogger
+
+extension BraveBookmarksAPI {
+
+    // MARK: Internal
+    
+    // Returns the last visited folder
+    // If no folder was visited, returns the mobile bookmarks folder
+    // If the root folder was visited, returns nil
+    func lastVisitedFolder() -> BookmarkNode? {
+        guard Preferences.General.showLastVisitedBookmarksFolder.value,
+              let nodeId = Preferences.Chromium.lastBookmarksFolderNodeId.value else {
+            // Default folder is the mobile node..
+            if let mobileNode = mobileNode {
+                return mobileNode
+            }
+            return nil
+        }
+        
+        // Display root folder instead of mobile node..
+        if nodeId == -1 {
+            return nil
+        }
+        
+        // Display last visited folder..
+        if let folderNode = getNodeById(nodeId),
+           folderNode.isVisible {
+            return folderNode
+        }
+        
+        // Default folder is the mobile node..
+        if let mobileNode = mobileNode {
+            return mobileNode
+        }
+        return nil
+    }
+    
+    func lastFolderPath() -> [BookmarkNode] {
+        if Preferences.General.showLastVisitedBookmarksFolder.value,
+           let nodeId = Preferences.Chromium.lastBookmarksFolderNodeId.value,
+           var folderNode = getNodeById(nodeId),
+           folderNode.isVisible {
+            
+            // We don't ever display the root node
+            // It is the mother of all nodes
+            let rootNodeGuid = rootNode?.guid
+            
+            var nodes = [BookmarkNode]()
+            nodes.append(folderNode)
+            
+            while true {
+                if let parent = folderNode.parent, parent.isVisible, parent.guid != rootNodeGuid {
+                    nodes.append(parent)
+                    folderNode = parent
+                    continue
+                }
+                break
+            }
+            return nodes.map({ $0 }).reversed()
+        }
+        
+        // Default folder is the mobile node..
+        if let mobileNode = mobileNode {
+            return [mobileNode]
+        }
+        
+        return []
+    }
+    
+    func mobileNode() -> BookmarkNode? {
+        if let node = mobileNode {
+            return node
+        }
+        return nil
+    }
+    
+    func fetchParent(_ bookmarkItem: BookmarkNode?) -> BookmarkNode? {
+        guard let bookmarkItem = bookmarkItem else {
+            return nil
+        }
+        
+        if let parent = bookmarkItem.parent {
+            // Return nil if the parent is the ROOT node
+            // because AddEditBookmarkTableViewController.sortFolders
+            // sorts root folders by having a nil parent.
+            // If that code changes, we should change here to match.
+            if bookmarkItem.parent?.guid != rootNode?.guid {
+                return parent
+            }
```

Comment:
This can be part of the enclosing if statement since there are no other branches

Reply:
Same as https://github.com/brave/brave-ios/pull/4024#discussion_r710328713 Fix at https://github.com/brave/brave-ios/pull/4024/commits/afa0c0001e63c7804bd48cfdf71bfc66d730557b